registry/health/health.go

```diff
@@ -0,0 +1,99 @@
+package health
+
+import (
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var (
+	mutex      sync.RWMutex
+	statusKeys = make(map[string]Status)
+)
+
+// Check set the interface for a Health Check
+type Check func() Status
```

Comment:
We can replace this with an interface:

```go
type Checker interface {
    // Check returns nil if the service is okay.
    Check() error
}
```

Note that we've replaced the explicit status type with an `error` interface. We can hide all sorts of data behind this or just export a `Status` struct that implements `error`. From this, we can simplify the register function:

```go
// Register associates the checker with the provided name. If a duplicate 
// name is registered, function panics.
func Register(name string, check Checker)
```

This allows us to build up different behaviors for different kinds of checks. For example, here is a settable `Checker` provided by the `health` package:

```go
// Updater implements a health check that is explicitly set.
type Updater interface {
    Checker

    // Update updates the current status of the health check.
    Update(status error)
}

// NewStatusUpdater returns an explicitly settable status check.
func NewStatusUpdater() Updater
```

Using the `Updater`, we can build up a periodic `Checker`:

```go
// PeriodicCheck wraps check to be called periodically.
func PeriodicCheck(check Checker, period time.Duration) Checker

// ... later when registering
Register("periodic_check", PeriodCheck(expensive, 10 * time.Second))
```

Internally, the `PeriodicCheck` function just uses an `Updater` with a goroutine loop.

This design supports the following use cases with a minimal abstraction:

1. Completely fresh implementation of `Check` interface (general case).
2. Directly settable statuses, where they make sense (explicit set case).
3. Periodic checks that may be expensive (primary use case).

I hope you don't consider this too much of a re-design but it makes the package more flexible. The checks themselves speak the familiar `error` interface and status code become an internal detail. Degraded or warning states can be communicated by an extended error interface:

```go
type Degraded interface {
    error
    Degraded() string
}
```

Although, such a definition isn't strictly necessary.